File: 2022/6.py
def prepare_data(file="input.txt"):
    with open(file) as file:
        lines = [line.strip() for line in file.readlines()]

    data = {"line": lines[0]}
    return data


def solve_n(line, n):
    chars = []
    for i, c in enumerate(line):
        if len(chars) == n:
            chars = chars[1:]
        if c in chars:
            chars += [c]
        else:
            if len(set(chars)) == n - 1:
                return i + 1
            chars += [c]
    return None


def solve1(line, **kwargs):
    return solve_n(line, 4)


def solve2(line, **kwargs):
    return solve_n(line, 14)

File: 2016/6.rs
use std::collections::HashMap;
use std::fs;

fn main() {
    let contents = fs::read_to_string("input.txt").expect("Error");
    let lines: Vec<&str> = contents.split("\n").collect();

    let mut counters = vec![
        HashMap::new(),
        HashMap::new(),
        HashMap::new(),
        HashMap::new(),
        HashMap::new(),
        HashMap::new(),
        HashMap::new(),
        HashMap::new(),
    ];
    let mut ans1 = String::new();
    let mut ans2 = String::new();

    for line in &lines {
        for (i, c) in line.chars().enumerate() {
            let counter = &mut counters[i];
            *counter.entry(c).or_insert(0) += 1;
        }
    }

    for i in 0..8 {
        let counter = &counters[i];
        let mut max = 0;
        let mut max_char = ' ';

        let mut min = std::usize::MAX;
        let mut min_char = ' ';

        for (c, count) in counter {
            if count > &max {
                max = *count;
                max_char = *c;
            }
            if count < &min {
                min = *count;
                min_char = *c;
            }
        }
        ans1 = format!("{}{}", ans1, max_char);
        ans2 = format!("{}{}", ans2, min_char);
    }

    println!("Part 1 answer:");
    println!("{}", ans1);

    println!("Part 2 answer:");
    println!("{}", ans2);
}
